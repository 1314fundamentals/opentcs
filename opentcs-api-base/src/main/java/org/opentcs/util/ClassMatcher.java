// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

import java.io.Serializable;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Predicate;

/**
 * Checks whether the input argument is assignable to any of a given set of classes.
 */
public class ClassMatcher
    implements
      Predicate<Object>,
      Serializable {

  /**
   * The set of classes to evaluate incoming events to.
   */
  private final Set<Class<?>> clazzes = new HashSet<>();

  /**
   * Creates a new instance.
   *
   * @param events The set of classes to evaluate incoming events to.
   */
  public ClassMatcher(Class<?>... events) {
    clazzes.addAll(Arrays.asList(events));
  }

  @Override
  public boolean test(Object object) {
    return clazzes.stream().anyMatch(clazz -> clazz.isAssignableFrom(object.getClass()));
  }

  @Override
  public Predicate<Object> negate() {
    return new ClassMatcher() {
      @Override
      public boolean test(Object t) {
        return !ClassMatcher.this.test(t);
      }
    };
  }
}
