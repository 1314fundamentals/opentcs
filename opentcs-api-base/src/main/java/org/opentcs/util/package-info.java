// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
/**
 * A collection of classes providing miscellaneous functions that are used
 * throughout openTCS.
 */
package org.opentcs.util;
