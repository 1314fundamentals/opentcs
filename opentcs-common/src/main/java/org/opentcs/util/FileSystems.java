// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

import static java.util.Objects.requireNonNull;

import java.io.File;

/**
 * This class provides helper methods for working with file systems.
 */
public final class FileSystems {

  /**
   * Prevents creation of instances.
   */
  private FileSystems() {
  }

  /**
   * Recursively deletes a given file/directory.
   *
   * @param target The file/directory to be deleted recursively.
   * @return <code>true</code> if deleting the target was successful, else
   * <code>false</code>.
   */
  public static boolean deleteRecursively(File target) {
    requireNonNull(target, "target");

    // If the target is a directory, remove its contents first.
    if (target.isDirectory()) {
      File[] entries = target.listFiles();
      for (File curEntry : entries) {
        boolean successful;

        if (curEntry.isDirectory()) {
          successful = deleteRecursively(curEntry);
        }
        else {
          successful = curEntry.delete();
        }

        if (!successful) {
          return false;
        }
      }
    }

    return target.delete();
  }
}
