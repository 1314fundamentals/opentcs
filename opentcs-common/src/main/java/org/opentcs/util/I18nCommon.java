// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

/**
 * Defines constants regarding internationalization.
 */
public interface I18nCommon {

  /**
   * The path to the project's resource bundle.
   */
  String BUNDLE_PATH = "i18n/org/opentcs/common/Bundle";
}
