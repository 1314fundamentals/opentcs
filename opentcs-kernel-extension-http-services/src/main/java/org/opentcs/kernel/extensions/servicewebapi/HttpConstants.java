// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

/**
 * Defines some HTTP-related constants.
 */
public class HttpConstants {

  /**
   * Name of the header that is expected to contain the API access keys.
   */
  public static final String HEADER_NAME_ACCESS_KEY = "X-Api-Access-Key";
  /**
   * Content type for plain text.
   */
  public static final String CONTENT_TYPE_TEXT_PLAIN_UTF8 = "text/plain; charset=utf-8";
  /**
   * Content type for JSON structures.
   */
  public static final String CONTENT_TYPE_APPLICATION_JSON_UTF8 = "application/json; charset=utf-8";

  /**
   * Prevents instantiation.
   */
  private HttpConstants() {
  }
}
