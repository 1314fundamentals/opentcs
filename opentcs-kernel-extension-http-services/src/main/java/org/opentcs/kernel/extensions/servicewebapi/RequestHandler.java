// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import io.javalin.apibuilder.EndpointGroup;
import org.opentcs.components.Lifecycle;

/**
 * A request handler.
 */
public interface RequestHandler
    extends
      Lifecycle {

  /**
   * Creates the handler's routes for the web service.
   */
  EndpointGroup createRoutes();
}
