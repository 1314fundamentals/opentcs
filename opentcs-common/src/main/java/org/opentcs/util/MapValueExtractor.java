// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

import static java.util.Objects.requireNonNull;

import jakarta.annotation.Nonnull;
import java.util.Map;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Provides utility methods for extracting values from {@link Map}s.
 */
public class MapValueExtractor {

  private static final Logger LOG = LoggerFactory.getLogger(MapValueExtractor.class);

  /**
   * Creates a new instance.
   */
  public MapValueExtractor() {
  }

  /**
   * Extracts a string value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @return An {@link Optional} containing the extracted value as a string, or an empty
   * {@link Optional} if the map does not contain an entry with the specified key.
   */
  public Optional<String> extractString(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");

    return Optional.ofNullable(map.get(key));
  }

  /**
   * Extracts an integer value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @return An {@link Optional} containing the extracted value parsed as an integer, or an empty
   * {@link Optional} if the map does not contain an entry with the specified key, or it's value
   * cannot be parsed.
   */
  public Optional<Integer> extractInteger(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");

    return Optional.ofNullable(map.get(key))
        .map(value -> {
          try {
            return Integer.valueOf(value);
          }
          catch (NumberFormatException e) {
            LOG.warn("Could not parse integer value for key '{}': {}", key, value);
            return null;
          }
        });
  }

  /**
   * Extracts a long value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @return An {@link Optional} containing the extracted value parsed as a long, or an empty
   * {@link Optional} if the map does not contain an entry with the specified key, or it's value
   * cannot be parsed.
   */
  public Optional<Long> extractLong(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");

    return Optional.ofNullable(map.get(key))
        .map(value -> {
          try {
            return Long.valueOf(value);
          }
          catch (NumberFormatException e) {
            LOG.warn("Could not parse long value for key '{}': {}", key, value);
            return null;
          }
        });
  }

  /**
   * Extracts a double value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @return An {@link Optional} containing the extracted value parsed as a double, or an empty
   * {@link Optional} if the map does not contain an entry with the specified key, or it's value
   * cannot be parsed.
   */
  public Optional<Double> extractDouble(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");

    return Optional.ofNullable(map.get(key))
        .map(value -> {
          try {
            return Double.valueOf(value);
          }
          catch (NumberFormatException e) {
            LOG.warn("Could not parse double value for key '{}': {}", key, value);
            return null;
          }
        });
  }

  /**
   * Extracts a boolean value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @return An {@link Optional} containing the extracted value parsed as a boolean, or an empty
   * {@link Optional} if the map does not contain an entry with the specified key, or it's value
   * cannot be parsed.
   */
  public Optional<Boolean> extractBoolean(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");

    return Optional.ofNullable(map.get(key))
        .map(Boolean::valueOf);
  }

  /**
   * Extracts an enum value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @param enumClass The class of the enum to extract.
   * @return An {@link Optional} containing the extracted value mapped to an enum of the
   * specified class, or an empty {@link Optional} if the map does not contain an entry with the
   * specified key, or it's value cannot be mapped.
   * @param <E> The type of the enum.
   */
  public <E extends Enum<E>> Optional<E> extractEnum(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map,
      @Nonnull
      Class<E> enumClass
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");
    requireNonNull(enumClass, "enumClass");

    return Optional.ofNullable(map.get(key))
        .map(value -> {
          try {
            return Enum.valueOf(enumClass, value);
          }
          catch (IllegalArgumentException e) {
            LOG.warn("Could not map enum value for key '{}': {}", key, value);
            return null;
          }
        });
  }
}
