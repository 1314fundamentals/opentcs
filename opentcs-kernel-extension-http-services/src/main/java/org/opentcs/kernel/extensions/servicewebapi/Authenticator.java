// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import static java.util.Objects.requireNonNull;

import com.google.common.base.Strings;
import io.javalin.http.Context;
import jakarta.inject.Inject;
import java.util.Objects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Authenticates incoming requests.
 */
public class Authenticator {

  /**
   * This class's logger.
   */
  private static final Logger LOG = LoggerFactory.getLogger(Authenticator.class);
  /**
   * Defines the required access rules.
   */
  private final ServiceWebApiConfiguration configuration;

  /**
   * Creates a new instance.
   *
   * @param configuration Defines the required access rules.
   */
  @Inject
  public Authenticator(ServiceWebApiConfiguration configuration) {
    this.configuration = requireNonNull(configuration, "configuration");
  }

  /**
   * Checks whether authentication is required and the given request is authenticated.
   *
   * @param context The context to be checked.
   * @return <code>true</code> if, and only if, authentication is required and the given request is
   * authenticated.
   */
  public boolean isAuthenticated(Context context) {
    requireNonNull(context, "context");

    String requestAccessKey = context.header(HttpConstants.HEADER_NAME_ACCESS_KEY);
    LOG.debug(
        "Provided access key in header is '{}', required value is '{}'",
        requestAccessKey,
        configuration.accessKey()
    );

    // Any empty access key indicates authentication is not required.
    if (Strings.isNullOrEmpty(configuration.accessKey())) {
      LOG.debug("No access key, authentication not required.");
      return true;
    }

    return Objects.equals(requestAccessKey, configuration.accessKey());
  }

}
