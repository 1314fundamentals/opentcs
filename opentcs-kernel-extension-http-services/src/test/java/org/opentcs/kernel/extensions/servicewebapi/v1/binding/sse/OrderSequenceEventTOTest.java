// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;

/**
 * Tests for {@link OrderSequenceEventTO}.
 */
class OrderSequenceEventTOTest {

  private JsonBinder jsonBinder;

  @BeforeEach
  void setUp() {
    jsonBinder = new JsonBinder();
  }

  @Test
  void jsonSample() {
    OrderSequenceEventTO to = new OrderSequenceEventTO(
        createOrderSequence(true),
        createOrderSequence(false)
    );

    Approvals.verify(jsonBinder.toJson(to));
  }

  private OrderSequenceEventTO.OrderSequenceTO createOrderSequence(boolean finished) {
    Instant time = Instant.parse("2025-07-22T12:03:00Z");
    return new OrderSequenceEventTO.OrderSequenceTO()
        .setName("order-sequence-1")
        .setProperties(new TreeMap<>(Map.of("key-1", "value-1", "key-2", "value-2")))
        .setHistory(
            new ObjectHistoryTO().setEntries(
                List.of(
                    new ObjectHistoryTO.ObjectHistoryEntryTO()
                        .setEventCode("event-code-1")
                        .setTimestamp(time)
                        .setSupplements(List.of("supplement-1", "supplement-2"))
                )
            )
        )
        .setType("type-1")
        .setOrders(List.of("order-1", "order-2"))
        .setFinishedIndex(4)
        .setComplete(true)
        .setFinished(finished)
        .setFailureFatal(true)
        .setIntendedVehicle("vehicle-1")
        .setProcessingVehicle("vehicle-2")
        .setCreationTime(time)
        .setFinishedTime(time.plus(1, ChronoUnit.HOURS));
  }
}
