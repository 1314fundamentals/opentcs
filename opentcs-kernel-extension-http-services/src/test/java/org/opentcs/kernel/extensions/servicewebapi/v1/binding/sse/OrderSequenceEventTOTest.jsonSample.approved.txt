{
  "currentObjectState" : {
    "name" : "order-sequence-1",
    "properties" : {
      "key-1" : "value-1",
      "key-2" : "value-2"
    },
    "history" : {
      "entries" : [ {
        "timestamp" : "2025-07-22T12:03:00Z",
        "eventCode" : "event-code-1",
        "supplements" : [ "supplement-1", "supplement-2" ]
      } ]
    },
    "type" : "type-1",
    "orders" : [ "order-1", "order-2" ],
    "finishedIndex" : 4,
    "complete" : true,
    "finished" : true,
    "failureFatal" : true,
    "intendedVehicle" : "vehicle-1",
    "processingVehicle" : "vehicle-2",
    "creationTime" : "2025-07-22T12:03:00Z",
    "finishedTime" : "2025-07-22T13:03:00Z"
  },
  "previousObjectState" : {
    "name" : "order-sequence-1",
    "properties" : {
      "key-1" : "value-1",
      "key-2" : "value-2"
    },
    "history" : {
      "entries" : [ {
        "timestamp" : "2025-07-22T12:03:00Z",
        "eventCode" : "event-code-1",
        "supplements" : [ "supplement-1", "supplement-2" ]
      } ]
    },
    "type" : "type-1",
    "orders" : [ "order-1", "order-2" ],
    "finishedIndex" : 4,
    "complete" : true,
    "finished" : false,
    "failureFatal" : true,
    "intendedVehicle" : "vehicle-1",
    "processingVehicle" : "vehicle-2",
    "creationTime" : "2025-07-22T12:03:00Z",
    "finishedTime" : "2025-07-22T13:03:00Z"
  }
}